internal/kubernetes/pod_introspector.go

```diff
@@ -0,0 +1,64 @@
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NoopPodIntrospector returns no imagePullSecrets; useful when introspection is not applicable.
+type NoopPodIntrospector struct{}
+
+func (*NoopPodIntrospector) GetImagePullSecrets(ctx context.Context) []corev1.LocalObjectReference {
+	return nil
+}
+
+func NewNoopPodIntrospector() *NoopPodIntrospector {
+	return &NoopPodIntrospector{}
+}
+
+// NewPodIntrospector creates a new PodIntrospector instance
+func NewPodIntrospector(client client.Client) (*PodIntrospector, error) {
+	podName := os.Getenv("POD_NAME")
+	podNamespace := os.Getenv("NAMESPACE")
+	if podName == "" || podNamespace == "" {
+		return nil, errors.New("POD_NAME or NAMESPACE environment variables not set, cannot introspect imagePullSecrets")
+	}
+
+	pi := &PodIntrospector{
+		client: client,
+		namespacedName: types.NamespacedName{
+			Name:      podName,
+			Namespace: podNamespace,
+		},
+	}
+
+	return pi, nil
+}
+
+// PodIntrospectorInterface provides access to information about the current pod
+type PodIntrospectorInterface interface {
```

Comment:
I feel like it's a bit odd to call out something directly as an "Interface" in the name (even though I know it's a thing we do in some places).

We could have this be the `PodIntrospector` then have the implementations be `NoopPodIntrospector` and `KubePodIntrospector` (or whatever better name you can come up with). This kind of aligns more with thinking of interfaces in terms of inheritance.